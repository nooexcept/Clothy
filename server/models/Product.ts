import mongoose from 'mongoose'

export interface Product extends mongoose.Document {
    name: string
    url: string
    description: string
    stock: number
    price: number
    cod: string
    images: Array<string>
}

const productSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true,
        unique: true,
    },
    url: {
        type: String,
        required: true,
        unique: true,
    },
    description: {
        type: String,
        required: true,
    },
    stock: {
        type: Number,
        required: true,
    },
    price: {
        type: Number,
        required: true,
    },
    cod: {
        type: String,
        required: true,
    },
    images: {
        type: Array,
        required: true,
    },
})

export default mongoose.model<Product>('Product', productSchema)
