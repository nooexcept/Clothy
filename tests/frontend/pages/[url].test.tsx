import React from 'react'
import { ShallowWrapper } from 'enzyme'
import { createShallow } from '@material-ui/core/test-utils'
import Product, {
    getStaticPaths,
    getStaticProps,
} from '../../../pages/products/[url]'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { ParsedUrlQuery } from 'querystring'

const mockedProducts = [
    {
        name: 'Ultra chill jumpers',
        url: 'ultra-chill-jumpers',
        description: 'If you wanna look chill, you need ultra chill jumpers',
        stock: 6,
        price: 80,
        cod: 'JUMPERS002',
        images: ['/photo-of-woman-standing-on-staircase-3120339.jpg'],
    },
]

let shallow: ReturnType<typeof createShallow>
let wrapper: ShallowWrapper
const mock = new MockAdapter(axios)

beforeEach(() => {
    shallow = createShallow()
    wrapper = shallow(<Product product={mockedProducts[0]} />)
})

afterEach(() => {
    mock.reset()
})

describe('<Product /> rendering', () => {
    it('should render without problems', () => {
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<Product /> data fetching', () => {
    it('should return paths from db, in getStaticPaths', async () => {
        mock.onGet(
            `http://localhost:${process.env.EXPRESS_PORT}/api/products/`
        ).reply(200, mockedProducts)

        const res = await getStaticPaths()

        expect(
            ((res.paths[0] as unknown) as ParsedUrlQuery & {
                params: { url: string }
            }).params.url
        ).toEqual(mockedProducts[0].url)
    })

    it('should return a product from db when specifying its url, in getStaticProps', async () => {
        mock.onGet(
            `http://localhost:${process.env.EXPRESS_PORT}/api/products/${mockedProducts[0].url}`
        ).reply(200, mockedProducts[0])

        const res = await getStaticProps({
            params: { url: mockedProducts[0].url },
        })
        expect(res.props.product).toEqual(mockedProducts[0])
    })
})
