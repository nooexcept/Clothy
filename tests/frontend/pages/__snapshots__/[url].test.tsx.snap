// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> rendering should render without problems 1`] = `
<PublicLayout>
  <div
    className="makeStyles-root-1"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={5}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={8}
      >
        <Styled(MuiBox)
          borderRight={1}
          className="makeStyles-picWrapper-7"
          height="100%"
        >
          <img
            className="makeStyles-mainPic-6"
            src="/photo-of-woman-standing-on-staircase-3120339.jpg"
          />
        </Styled(MuiBox)>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <Styled(MuiBox)
          m={2}
        >
          <Styled(MuiBox)
            ml={1}
          >
            <WithStyles(ForwardRef(Typography))
              gutterBottom={true}
              variant="caption"
            >
              JUMPERS002
            </WithStyles(ForwardRef(Typography))>
          </Styled(MuiBox)>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-titleTypo-5"
            component="h3"
            gutterBottom={true}
            variant="h2"
          >
            Ultra chill jumpers
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-priceTypo-4"
            gutterBottom={true}
            variant="h6"
          >
            $80
          </WithStyles(ForwardRef(Typography))>
        </Styled(MuiBox)>
        <Styled(MuiBox)
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
            }
          }
        >
          <AnnouncementIcon
            color="secondary"
          />
          <WithStyles(ForwardRef(Typography))
            color="secondary"
            display="inline"
            gutterBottom={true}
            variant="body1"
          >
            Almost out of stock
          </WithStyles(ForwardRef(Typography))>
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Button))
          classes={
            Object {
              "label": "makeStyles-buyBtnLabel-3",
              "root": "makeStyles-buyBtn-2",
            }
          }
        >
          Add to cart
        </WithStyles(ForwardRef(Button))>
        <Styled(MuiBox)
          mb={3}
          mt={3}
        >
          <WithStyles(ForwardRef(Divider)) />
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="subtitle1"
        >
          If you wanna look chill, you need ultra chill jumpers
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </div>
</PublicLayout>
`;
